File: src/main/java/no/nav/fo/veilarbregistrering/arbeidssoker/adapter/baseclient/LogInterceptor.java
package no.nav.fo.veilarbregistrering.arbeidssoker.adapter.baseclient;

import no.nav.log.LogFilter;
import no.nav.log.MDCConstants;
import no.nav.sbl.util.EnvironmentUtils;
import no.nav.sbl.util.StringUtils;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

import java.io.IOException;

import static java.lang.String.format;
import static java.util.Arrays.stream;
import static no.nav.log.LogFilter.NAV_CALL_ID_HEADER_NAMES;

public class LogInterceptor implements Interceptor {

    private static final Logger LOG = LoggerFactory.getLogger(LogInterceptor.class);

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();
        Request.Builder requestBuilder = original.newBuilder();

        StringUtils.of(MDC.get(MDCConstants.MDC_CALL_ID))
                .ifPresent(callId -> stream(NAV_CALL_ID_HEADER_NAMES)
                        .forEach(headerName-> requestBuilder.header(headerName, callId)));

        EnvironmentUtils.getApplicationName().ifPresent(applicationName -> requestBuilder.header(LogFilter.CONSUMER_ID_HEADER_NAME, applicationName));

        Request request = requestBuilder
                .method(original.method(), original.body())
                .build();

        try {
            Response response = chain.proceed(request);
            LOG.info(format("%d %s %s", response.code(), request.method(), request.url()));
            return response;
        } catch (Exception exception) {
            LOG.error(format("Request failed: %s %s", request.method(), request.url()), exception);
            throw exception;
        }
    }

}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/http/LogInterceptor.kt
package no.nav.fo.veilarbregistrering.http

import no.nav.common.log.LogFilter
import no.nav.common.log.MDCConstants
import no.nav.common.utils.IdUtils
import no.nav.fo.veilarbregistrering.config.applicationNameOrNull
import no.nav.fo.veilarbregistrering.log.logger
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import org.slf4j.MDC
import java.math.RoundingMode
import java.text.DecimalFormat

class LogInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val startTime = System.nanoTime()
        val incomingRequest = chain.request()
        var throwable: Throwable? = null
        var response: Response? = null
        val requestBuilder = incomingRequest.newBuilder()

        context.addCallIdHeaders(requestBuilder)
        context.addConsumerId(requestBuilder)

        val request = requestBuilder.method(incomingRequest.method(), incomingRequest.body()).build()

        try {
            response = chain.proceed(request)
            val delta = ((System.nanoTime() - startTime) / 1e6)
            logger.info("${response.code()} ${request.method()} ${request.url()} ${context.timeFormat.format(delta)}mS.")
        } catch (t: Throwable) {
            throwable = t
        }

        return when (throwable) {
            null -> response ?: throw IllegalStateException("Error in RequestFilter, missing response")
            else -> {
                throw throwable
            }
        }
    }
}

private val context = object {
    val NAV_CALL_ID_HEADER_NAMES = listOf("Nav-Call-Id", "Nav-CallId", "X-Correlation-Id")
    val timeFormat = DecimalFormat("#.##").also { it.roundingMode = RoundingMode.CEILING }

    fun addConsumerId(requestBuilder: Request.Builder) {
        applicationNameOrNull()?.let { applicationName: String ->
            requestBuilder.header(
                LogFilter.CONSUMER_ID_HEADER_NAME,
                applicationName
            )
        }
    }

    fun addCallIdHeaders(requestBuilder: Request.Builder) {
        val callId = MDC.get(MDCConstants.MDC_CALL_ID)
            ?: MDC.get(MDCConstants.MDC_JOB_ID)
            ?: IdUtils.generateId() // Generate a new call-id if it is missing from the MDC context

        NAV_CALL_ID_HEADER_NAMES.forEach { headerName: String ->
            requestBuilder.header(
                headerName,
                callId
            )
        }
    }
}